tools/tfdoc.py

```diff
@@ -14,274 +14,350 @@
-      self.in_progress = True
-      self.name = token_data
-    elif token_type == 'DESCRIPTION':
-      setattr(self, token_type.lower(), token_data)
-    elif token_type in ('DEFAULT', 'TYPE'):
-      self._start(token_type.lower(), token_data)
-    elif token_type == 'REST':
-      self._data.append(token_data)
-
-  def _close(self, strip=False):
-    if self._data_context:
-      data = self._data
-      if strip and '}' in data[-1]:
-        data = data[:-1]
-      setattr(self, self._data_context, ('\n'.join(data)).strip())
-
-  def _start(self, context, data):
-    if context == self._data_context or getattr(self, context):
-      self._data.append("%s = %s" % (context, data))
-      return
-    self._close()
-    self._data = [data]
-    self._data_context = context
-
-  def close(self):
-    self._close(strip=True)
-    return VariableData(self.name, self.description, self.type, self.default,
-                        self.default is None)
+VAR_RE_TYPE = re.compile(r'([\(\{\}\)])')
+VAR_TEMPLATE = ('default', 'description', 'type')
+
+
+File = collections.namedtuple('File', 'name description modules resources')
+Output = collections.namedtuple('Output',
+                                'name description sensitive consumers')
+Variable = collections.namedtuple(
+    'Variable', 'name description type default required source')
+
+
+# parsing functions
+
+
+def _extract_tags(body):
+  'Extract and return tfdocs tags from content.'
+  return {k: v for k, v in TAG_RE.findall(body)}
+
+
+def _parse(body, enum=VAR_ENUM, re=VAR_RE, template=VAR_TEMPLATE):
+  'Low-level parsing function for outputs and variables.'
+  item = context = None
+  for m in re.finditer(body):
+    token = enum(m.lastindex)
+    data = m.group(m.lastindex)
+    # print(token, m.groups())
+    if token == enum.OPEN:
+      item = {'name': data, 'tags': {}}
+      item.update({k: [] for k in template})
+      context = None
+    elif token == enum.CLOSE:
+      if item:
+        yield item
+      item = context = None
+    elif token == enum.ATTR_DATA:
+      if not item:
+        continue
+      context = m.group(m.lastindex-1)
+      item[context].append(data)
+    elif token == enum.SKIP:
+      context = token
+    elif token == enum.COMMENT:
+      if item and data.startswith('tfdoc:'):
+        k, v = data.split(' ', 1)
+        item['tags'][k[6:]] = v
+    elif token == enum.TXT:
+      if context and context != enum.SKIP:
+        item[context].append(data)
+
+
+def parse_files(basepath):
+  'Return a list of File named tuples in root module at basepath.'
+  for name in glob.glob(os.path.join(basepath, '*tf')):
+    try:
+      with open(name) as file:
+        body = file.read()
+    except (IOError, OSError) as e:
+      raise SystemExit(f'Cannot read file {name}: {e}')
+    shortname = os.path.basename(name)
+    tags = _extract_tags(body)
+    description = tags.get(
+        'file:description', FILE_DESC_DEFAULTS.get(shortname))
+    modules = set([
```

Comment:
no need for the list comprehension here. Just set(...) should be enough

Reply:
good point :)